### DISCUSSION TITLE
[@types/marked] Argument of type 'Renderer<never>' is not assignable to parameter of type 'RendererThis'

### DISCUSSION THREAD

**Discussion by jfcere** [DefinitelyTyped/DefinitelyTyped#58301] [Issues with a @types package] [answered]
### The Problem

Using marked 2.0 with TypeScript I was able to override the `link` token with `renderer` and call back the original `link` function on the prototype like this:

```typescript
import { Renderer } from 'marked';

const renderer = new Renderer();

renderer.link = (href: string, title: string, text: string) => {
  return Renderer.prototype.link.call(renderer, normalizeExternalUrl(href), title, text) as string;
};
```

After updating from marked 2.0 to 4.0 I am trying to migrate the code above but I am getting an error telling me that the parameter `this` is not of the expected type `RendererThis`.

![image](https://user-images.githubusercontent.com/6987084/149675272-2c8257cc-4db7-447a-b6c1-d1f0b9ac0d34.png)

> Note that the `normalizeExternalUrl` function return type is `string`, it is only to handle relative links

### Investigation

I took a look at the typings for @types/marked but `RendererThis` doesn't make much sense to me as if I put a breakpoint "inside" the `link` function and inspect it with Chrome DevTools `this` is shown as being of type `Renderer`...

![image](https://user-images.githubusercontent.com/6987084/150044645-af5d048e-5a71-4466-b938-4098897f9730.png)

### Proposed Change

I would greatly be in need of advice because my first impression would be to remove `RendererThis` to replace it by `Renderer` pretty much everywhere it is used... but surely I am missing something!

Any opinion on the subject is welcome!

[a comment by an automated account is not shown]

**Comment by UziTech** [comment]
The issue is that there are two ways to create a renderer in marked. Renderer class (`const renderer = new Renderer()`) in which case `this` is the renderer object, and custom extension renderer (`const extension = {name: 'test', renderer() {...}}`) in which case `this` is the `RendererThis` in the types.

I think the solution should be to fix marked and use `RendererThis` and `TokenizerThis` for all renderer/tokenizer functions.

**Reply by jfcere** [reply]
@UziTech how about using union type `Renderer | RendererThis` to type `this`?

For example...

```typescript
link(this: Renderer | RendererThis, href: string | null, title: string | null, text: string): string | T;
```

**Reply by UziTech** [reply]
that could work. @jfcere if you want to create a PR for that I would approve it. Make sure you add a test to make sure your code works.

**Reply by jfcere** [reply]
Sure, I'll do it this weekend.
Thanks for your input and your help on the subject!

I'll keep you posted.

**Reply by jfcere** [reply]
The PR has been submitted: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58495

**Comment by jfcere** [comment, marked as answer]
The PR to fix the typings has been merged, deployed as @types/marked 4.0.2 and tested.

Thanks again @UziTech 👍
